src/transaction/meta_store_handler.rs

```diff
@@ -0,0 +1,203 @@
+    executor: Arc<Exe>,
+    transaction_coordinator_id: u32,
+}
+
+impl<Exe: Executor> TransactionMetaStoreHandler<Exe> {
+    // Should we make these configurable?
+    const MAX_OP_RETRIES: u32 = 3;
+    const OP_MIN_BACKOFF: Duration = Duration::from_millis(10);
+    const OP_MAX_BACKOFF: Duration = Duration::from_millis(1000);
+
+    #[cfg_attr(feature = "telemetry", tracing::instrument(skip_all))]
+    pub async fn new(
+        client: Pulsar<Exe>,
+        connection_mgr: Arc<ConnectionManager<Exe>>,
+        connection: RwLock<Arc<Connection<Exe>>>,
+        executor: Arc<Exe>,
+        transaction_coordinator_id: u32,
+    ) -> Result<Self, Error> {
+        let handler = Self {
+            client,
+            connection_mgr,
+            connection,
+            executor,
+            transaction_coordinator_id,
+        };
+
+        handler.tc_client_connect().await?;
+
+        Ok(handler)
+    }
+
+    pub async fn reconnect(&self) -> Result<(), Error> {
+        // `get_tc_connection` will reconnect and swap the connection associated with the address
+        // if the connection is not healthy. If the topic lookup points to the same broker
+        // and the connection is healthy, this is effectively a no-op. In the case we discover
+        // the TC is on a different broker, we'll swap the connection to the new broker's address.
+        *self.connection.write().expect("poisoned lock") = super::get_tc_connection(
+            &self.client,
+            &self.connection_mgr,
+            self.transaction_coordinator_id,
+        )
+        .await?;
+
+        Ok(())
+    }
+
+    #[cfg_attr(feature = "telemetry", tracing::instrument(skip_all))]
+    async fn extract_with_retry<T, F, E, R, S>(&self, op: F, extract: E) -> Result<S, Error>
+    where
+        R: Future<Output = Result<T, ConnectionError>>,
+        F: Fn() -> R,
+        E: Fn(T) -> ((Option<i32>, Option<String>), S),
+    {
+        let mut current_retries = 0u32;
+        let mut current_backoff;
+
+        loop {
+            let res = op().await?;
+            let ((code, message), extracted) = extract(res);
+
+            if let Some(err) = code {
+                let server_error = server_error(err);
+
+                match server_error {
+                    // TC Not Found errors shouldn't bubble up to the user before
+                    // we've tried reconnecting a few times
+                    error @ Some(ServerError::TransactionCoordinatorNotFound) => {
+                        if current_retries >= Self::MAX_OP_RETRIES {
+                            return Err(ConnectionError::PulsarError(error, message).into());
+                        }
+
+                        let jitter = rand::thread_rng().gen_range(0..10);
+                        current_backoff = std::cmp::min(
+                            Self::OP_MAX_BACKOFF * 2u32.saturating_pow(current_retries),
+                            Self::OP_MAX_BACKOFF,
```

Comment:
Why use `cmp::min` here? `OP_MAX_BACKOFF` should always be the smaller one. It looks like you should use `OP_MIN_BACKOFF * 2u32.saturating_pow(current_retries)`.

BTW, could you reuse the `operation_retry_parameters` in `Pulsar::new` instead of hard coding backoff parameters? And it would be better to abstract a `Backoff` class like Java so that we can reuse the logic in `connect_inner` rather than rewriting the same logic again

Reply:
> Why use `cmp::min` here? `OP_MAX_BACKOFF` should always be the smaller one. It looks like you should use `OP_MIN_BACKOFF * 2u32.saturating_pow(current_retries)`.

Thanks for catching that! Updated to use `OP_MIN_BACKOFF` there.

> BTW, could you reuse the `operation_retry_parameters` in `Pulsar::new` instead of hard coding backoff parameters?

The same `operation_retry_parameters` as declared in `Pulsar::new` should already be getting used here because these operations rely on `ConnectionSender::send_message` .

The retry logic here is solely for `TransactionCoordinatorNotFound` errors. I think it's possible to make the backoff parameters here configurable, but they should be different from `operation_retry_parameters`.

> And it would be better to abstract a `Backoff` class like Java so that we can reuse the logic in `connect_inner` rather than rewriting the same logic again

Agreed, but I think maybe we can make another issue for this? May not be a great idea to increase the scope of this PR since it's already quite large.